src/mono/System.Private.CoreLib/src/System/Collections/Generic/ArraySortHelper.Mono.cs

```diff
@@ -1,15 +1,77 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
+using System.Runtime.CompilerServices;
+using System.Diagnostics.CodeAnalysis;
+
 namespace System.Collections.Generic
 {
-    internal partial class ArraySortHelper<T>
+    internal interface IArraySortHelper<TKey>
+    {
+        void Sort(Span<TKey> keys, IComparer<TKey>? comparer);
+        int BinarySearch(TKey[] keys, int index, int length, TKey value, IComparer<TKey>? comparer);
+    }
+
+    internal sealed partial class ArraySortHelper<T>
+        : IArraySortHelper<T>
+    {
+        private static readonly IArraySortHelper<T> s_defaultArraySortHelper = CreateArraySortHelper();
+
+        public static IArraySortHelper<T> Default => s_defaultArraySortHelper;
+
+        [DynamicDependency("#ctor", typeof(GenericArraySortHelper<>))]
+        private static IArraySortHelper<T> CreateArraySortHelper()
+        {
+            IArraySortHelper<T> defaultArraySortHelper;
+
+            if (typeof(IComparable<T>).IsAssignableFrom(typeof(T)))
+            {
+                defaultArraySortHelper = (IArraySortHelper<T>)RuntimeType.CreateInstanceForAnotherGenericParameter((RuntimeType)typeof(GenericArraySortHelper<>), (RuntimeType)typeof(T));
+            }
+            else
+            {
+                defaultArraySortHelper = new ArraySortHelper<T>();
+            }
+            return defaultArraySortHelper;
+        }
+    }
+
+    internal sealed partial class GenericArraySortHelper<T>
+        : IArraySortHelper<T>
     {
-        public static ArraySortHelper<T> Default { get; } = new ArraySortHelper<T>();
     }
 
-    internal partial class ArraySortHelper<TKey, TValue>
+    internal interface IArraySortHelper<TKey, TValue>
+    {
+        void Sort(Span<TKey> keys, Span<TValue> values, IComparer<TKey>? comparer);
+    }
+
+    internal sealed partial class ArraySortHelper<TKey, TValue>
+        : IArraySortHelper<TKey, TValue>
+    {
+        private static readonly IArraySortHelper<TKey, TValue> s_defaultArraySortHelper = CreateArraySortHelper();
+
+        public static IArraySortHelper<TKey, TValue> Default => s_defaultArraySortHelper;
+
+        [DynamicDependency("#ctor", typeof(GenericArraySortHelper<,>))]
```

Comment:
Is this needed? `CreateInstanceForAnotherGenericParameter` has a `[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicParameterlessConstructor)]` annotation on the first parameter. It feels like a illink bug if the ctor is not kept.

Reply:
Its a copy of the corresponding file in coreclr:
https://github.com/dotnet/runtime/blob/main/src/coreclr/System.Private.CoreLib/src/System/Collections/Generic/ArraySortHelper.CoreCLR.cs

Reply:
Deleting that in #71369.